client/my-sites/themes/controller.js

```diff
@@ -1,97 +1,153 @@
+			'Themes'
 		);
-	},
+	};
 
-	details: function( context ) {
-		const user = getCurrentUser( context.store.getState() );
+	return {
+		title,
+		tier,
+		search: context.query.s,
+		trackScrollPage: boundTrackScrollPage,
+		runClientAnalytics
+	};
+}
 
-		const Head = user
-			? require( 'layout/head' )
-			: require( 'my-sites/themes/head' );
+function createElement( ThemesComponent, Head, store, props ) {
+	return(
+		<ReduxProvider store={ store }>
+			<Head title={ props.title } tier={ props.tier || 'all' }>
+				<ThemesComponent { ...omit( props, [ 'title', 'runClientAnalytics' ] ) } />
+				<ClientSideEffects>
+					{ props.runClientAnalytics }
+				</ClientSideEffects>
+			</Head>
+		</ReduxProvider>
+	);
+};
 
-		const element = (
-			<ReduxProvider store={ context.store } >
-				<Head title="Something Theme — WordPress.com" isSheet>
-					<ThemeSheetComponent themeSlug={ context.params.slug } />
-				</Head>
-			</ReduxProvider>
-		);
+export function singleSite( context, next ) {
+	const Head = require( 'layout/head' );
+	const { site_id: siteId } = context.params;
+	const props = getProps( context );
+
+	props.key = siteId;
+	props.siteId = siteId;
+
+	context.store.dispatch( setSection( 'design', {
+		hasSidebar: true,
+		isFullScreen: false
+	} ) );
+
+	context.primary = createElement( SingleSiteComponent, Head, context.store, props );
+	next();
+}
 
-		// FIXME: temporary hack until we have a proper isomorphic, one tree routing solution. Do NOT do this!
-		const sheetsDomElement = document.getElementsByClassName( 'themes__sheet' )[0];
-		if ( ! sheetsDomElement ) {
-			ReactDom.render( element, document.getElementById( 'primary' ) );
-		}
+export function multiSite( context, next ) {
+	const Head = require( 'layout/head' );
+	const props = getProps( context );
 
-		return element;
+	context.store.dispatch( setSection( 'design', {
+		hasSidebar: true,
+		isFullScreen: false
+	} ) );
+
+	context.primary = createElement( MultiSiteComponent, Head, context.store, props );
+	next();
+}
+
+export function loggedOut( context, next ) {
+	const Head = require( 'my-sites/themes/head' );
+	const props = getProps( context );
+
+	context.store.dispatch( setSection( 'design', {
+		hasSidebar: false,
+		isFullScreen: false
+	} ) );
+
+	context.primary = createElement( LoggedOutComponent, Head, context.store, props );
+	next();
+}
+
+export function details( context, next ) {
+	const user = getCurrentUser( context.store.getState() );
+	const Head = user
+		? require( 'layout/head' )
+		: require( 'my-sites/themes/head' );
+	const props = {
+		themeSlug: context.params.slug,
+		title: buildTitle(
+			i18n.translate( 'Theme Details', { textOnly: true } )
+		)
 	}
 
-};
+	context.store.dispatch( setSection( 'themes', {
+		hasSidebar: false,
+		isFullScreen: true
+	} ) );
 
-module.exports = controller;
+	// When we're logged in, we need to remove the sidebar.
+	ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
```

Comment:
> Since as you say the whole point of this PR is untangling middleware, do you think we should add that client-specific middleware to this PR already?

I'm inclined to make it a subsequent PR after this.

> I'm inclined to keep it separate.

Agreed.